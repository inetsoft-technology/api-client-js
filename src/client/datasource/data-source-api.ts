/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * DataSourceDescription contains a description of a data source.
 */
export interface DataSourceDescription {
   /**
    * The unique identifier of the data source.
    */
   id: string;

   /**
    * The name of the data source.
    */
   name: string;

   /**
    * The type of the data source.
    */
   type: string;
}

/**
 * DataSourceProperties contains the properties of a data source definition.
 */
export interface DataSourceProperties {
   /**
    * The unique identifier of the data source.
    */
   id: string;

   /**
    * The name of the data source.
    */
   name: string;

   /**
    * The type of the data source. Allowed values are 'jdbc', 'soap', 'text', 'xml', and 'tabular'.
    */
   type: string;
}

/**
 * JdbcDataSourceProperties contains the properties of a JDBC data source definition.
 */
export interface JdbcDataSourceProperties extends DataSourceProperties {
   /**
    * The JDBC URL of the database.
    */
   url: string;

   /**
    * The fully-qualified class name of the JDBC driver.
    */
   driver: string;

   /**
    * A flag that indicates if the database requires authentication.
    */
   requireLogin: boolean;

   /**
    * The user name used to authenticate with the database.
    */
   user?: string;

   /**
    * The password used to authenticate with the database.
    */
   password?: string;

   /**
    * The name of the default database.
    */
   defaultDatabase?: string;
}

/**
 * SoapDataSourceProperties contains the properties of a SOAP data source definition.
 */
export interface SoapDataSourceProperties extends DataSourceProperties {
   /**
    * The URL of the SOAP service.
    */
   url: string;

   /**
    * A flag that indicates if the SOAP service requires authentication.
    */
   requireLogin: boolean;

   /**
    * The user name used to authenticate with the SOAP service.
    */
   user?: string;

   /**
    * The password used to authenticate with the SOAP service.
    */
   password?: string;

   /**
    * The fully-qualified class name of the SOAP server object interface.
    */
   serverClass: string;
}

/**
 * TextDataSourceProperties contains the properties of a text data source definition.
 */
export interface TextDataSourceProperties extends DataSourceProperties {
   /**
    * The URL of the text file.
    */
   url: string;

   /**
    * The HTTP method used to retrieve the text file. This property is only used if the URL protocol
    * is http or https.
    */
   method?: string;
}

/**
 * XmlDataSourceProperties contains the properties of an XML data source definition.
 */
export interface XmlDataSourceProperties extends DataSourceProperties {
   /**
    * The URL of the XML file.
    */
   url: string;

   /**
    * The HTTP method used to retrieve the XML file. This property is only used if the URL protocol
    * is http or https.
    */
   method?: string;
}

/**
 * TabularDataSourceProperties contains the properties of a tabular data source definition.
 */
export interface TabularDataSourceProperties extends DataSourceProperties {
   /**
    * The tabular sub-type of the data source.
    */
   tabularType: string;
}

/**
 * DataSourceApi provides operations used to list and manage data sources.
 */
export interface DataSourceApi {
   /**
    * Gets the list of data sources.
    *
    * @return the data sources.
    */
   getDataSources(): Promise<DataSourceDescription[]>;

   /**
    * Gets the properties of a data source.
    *
    * @param name the name of the data source.
    *
    * @return the data source properties.
    */
   getDataSource(name: string): Promise<DataSourceProperties | null>;

   /**
    * Updates the properties of an existing data source.
    *
    * @param name       the name of the data source.
    * @param properties the modified data source properties.
    */
   updateDataSource(name: string, properties: DataSourceProperties): Promise<void>;

   /**
    * Deletes a data source.
    *
    * @param name  the name of the data source.
    * @param force a flag indicating if the data source should be deleted even if it is used by a
    *              query.
    */
   deleteDataSource(name: string, force?: boolean): Promise<void>;
}
