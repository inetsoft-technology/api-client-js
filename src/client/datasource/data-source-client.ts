/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { AbstractClient } from "../abstract-client";
import { HttpAdapter } from "../http-adapter";
import { DataSourceApi, DataSourceDescription, DataSourceProperties } from "./data-source-api";

/**
 * @hidden
 */
export class DataSourceClient extends AbstractClient implements DataSourceApi {

   constructor(adapter: HttpAdapter<any>, token: () => Promise<string>, baseUrl: string) {
      super(adapter, token, baseUrl);
   }

   public getDataSources(): Promise<DataSourceDescription[]> {
      const url = "api/public/dataSources";
      return this.doGet(url).then((list: any) => list.dataSources);
   }

   public getDataSource(name: string): Promise<DataSourceProperties | null> {
      return this.findDataSource(name);
   }

   public updateDataSource(name: string, properties: DataSourceProperties): Promise<void> {
      return this.getId(name).then(id => {
         if (id) {
            const url = `api/public/dataSources/${id}`;
            return this.doPut<void>(url, properties);
         }

         return Promise.reject(`Data source not found: '${name}'`);
      });
   }

   public deleteDataSource(name: string, force: boolean = false): Promise<void> {
      return this.getId(name).then(id => {
         if (id) {
            const url = `api/public/dataSources/${id}`;
            const params = {force: force ? "true" : "false"};
            return this.doDelete<void>(url, params);
         }

         return Promise.reject(`Data source not found: '${name}'`);
      });
   }

   private findDataSource(name: string): Promise<DataSourceProperties | null> {
      const url = "api/public/dataSources";
      const params = { name };
      return this.doGet<any>(url, params).then(list => {
         if (list.dataSources.length === 0) {
            return null;
         }

         return list.dataSources[0];
      });
   }

   private getId(name: string): Promise<string | null> {
      return this.findDataSource(name).then(ds => ds ? ds.id : null);
   }
}
