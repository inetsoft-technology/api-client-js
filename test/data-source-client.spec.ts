/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { DataSourceClient } from "../src/client/datasource/data-source-client";

describe("DataSourceClient", () => {
   let adapter: any;
   let client: DataSourceClient;

   const dataSources = [
      {
         id: "ds1",
         name: "Data source 1",
         type: "jdbc",
         url: "jdbc:test:ds1",
         driver: "com.example.Driver",
         requireLogin: true,
         user: "sa",
         password: "secret",
         defaultDatabase: "PUBLIC"
      },
      {
         id: "ds2",
         name: "Data source 2",
         type: "text",
         url: "http://localhost/data/example.csv",
         method: "get"
      }
   ];

   beforeEach(() => {
      adapter = {
         get: jest.fn(),
         post: jest.fn(),
         put: jest.fn(),
         delete: jest.fn(),
         map: jest.fn(),
         mapText: jest.fn(),
         mapBinary: jest.fn()
      };
      client = new DataSourceClient(adapter, () => Promise.resolve("secret"), "http://localhost:8080/sree/");
   });

   describe("when getDataSources() is called", () => {
      it("then a list of data sources is returned", () => {
         const list = { dataSources };

         adapter.get.mockImplementation(() => Promise.resolve(list));
         adapter.map.mockImplementation((response: Promise<any>) => response);

         return client.getDataSources().then(result => {
            expect(result).toMatchSnapshot();
            expect(adapter.get.mock.calls.length).toBe(1);
            expect(adapter.get.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/dataSources");
         });
      });
   });

   describe("when getDataSource() is called with a valid name", () => {
      it("then a data source is returned", () => {
         const list = { dataSources: [dataSources[0]] };
         adapter.get.mockReturnValueOnce(Promise.resolve(list));
         adapter.map.mockImplementation((response: Promise<any>) => response);

         return client.getDataSource("Data source 1").then(result => {
            expect(result).toMatchSnapshot();
            expect(adapter.get.mock.calls.length).toBe(1);
            expect(adapter.get.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/dataSources");
            expect(adapter.get.mock.calls[0][1].name).toBe("Data source 1");
         });
      });
   });

   describe("when getDataSource() is called with an invalid name", () => {
      it("then null is returned", () => {
         const list = { dataSources: [] };
         adapter.get.mockReturnValueOnce(Promise.resolve(list));
         adapter.map.mockImplementation((response: Promise<any>) => response);

         return client.getDataSource("Missing data source").then(result => {
            expect(result).toBeNull();
            expect(adapter.get.mock.calls.length).toBe(1);
            expect(adapter.get.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/dataSources");
            expect(adapter.get.mock.calls[0][1].name).toBe("Missing data source");
         });
      });
   });

   describe("when updateDataSource() is called with a valid name and is successful", () => {
      it("then the promise is resolved", () => {
         const list = { dataSources: [dataSources[0]] };
         adapter.get.mockReturnValueOnce(Promise.resolve(list));
         adapter.put.mockReturnValueOnce(Promise.resolve());
         adapter.map.mockImplementation((response: Promise<any>) => response);

         return client.updateDataSource("Data source 1", dataSources[1]).then(() => {
            expect(adapter.get.mock.calls.length).toBe(1);
            expect(adapter.get.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/dataSources");
            expect(adapter.get.mock.calls[0][1].name).toBe("Data source 1");
            expect(adapter.put.mock.calls.length).toBe(1);
            expect(adapter.put.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/dataSources/ds1");
            expect(adapter.put.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when updateDataSource() is called with a valid name and fails", () => {
      it("then the promise is rejected", () => {
         const list = { dataSources: [dataSources[0]] };
         adapter.get.mockReturnValueOnce(Promise.resolve(list));
         adapter.put.mockReturnValueOnce(Promise.reject("Test failure"));
         adapter.map.mockImplementation((response: Promise<any>) => response);

         expect.assertions(1);
         return expect(client.updateDataSource("Data source 1", dataSources[1])).rejects.toEqual("Test failure");
      });
   });

   describe("when updateDataSource() is called with an invalid name", () => {
      it("then the promise is rejected", () => {
         const list = { dataSources: [] };
         adapter.get.mockReturnValueOnce(Promise.resolve(list));
         adapter.map.mockImplementation((response: Promise<any>) => response);

         expect.assertions(1);
         return expect(client.updateDataSource("Missing data source", dataSources[1])).rejects.toEqual("Data source not found: 'Missing data source'");
      });
   });

   describe("when deleteDataSource() is called with a valid name and succeeds", () => {
      it("then the promise is resolved", () => {
         const list = { dataSources: [dataSources[0]] };
         adapter.get.mockReturnValueOnce(Promise.resolve(list));
         adapter.delete.mockReturnValueOnce(Promise.resolve());
         adapter.map.mockImplementation((response: Promise<any>) => response);

         return client.deleteDataSource("Data source 1").then(() => {
            expect(adapter.get.mock.calls.length).toBe(1);
            expect(adapter.get.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/dataSources");
            expect(adapter.get.mock.calls[0][1].name).toBe("Data source 1");
            expect(adapter.delete.mock.calls.length).toBe(1);
            expect(adapter.delete.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/dataSources/ds1");
         });
      });
   });

   describe("when deleteDataSource() is called with a valid name, the force parameter, and succeeds", () => {
      it("then the promise is resolved", () => {
         const list = { dataSources: [dataSources[0]] };
         adapter.get.mockReturnValueOnce(Promise.resolve(list));
         adapter.delete.mockReturnValueOnce(Promise.resolve());
         adapter.map.mockImplementation((response: Promise<any>) => response);

         return client.deleteDataSource("Data source 1", true).then(() => {
            expect(adapter.get.mock.calls.length).toBe(1);
            expect(adapter.get.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/dataSources");
            expect(adapter.get.mock.calls[0][1].name).toBe("Data source 1");
            expect(adapter.delete.mock.calls.length).toBe(1);
            expect(adapter.delete.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/dataSources/ds1");
            expect(adapter.delete.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when deleteDataSource() is called with a valid name and fails", () => {
      it("then the promise is rejected", () => {
         const list = {dataSources: [dataSources[0]]};
         adapter.get.mockReturnValueOnce(Promise.resolve(list));
         adapter.delete.mockReturnValueOnce(Promise.reject("Test failure"));
         adapter.map.mockImplementation((response: Promise<any>) => response);

         expect.assertions(1);
         return expect(client.deleteDataSource("Data source 1")).rejects.toEqual("Test failure");
      });
   });

   describe("when deleteDataSource() is called with an invalid name", () => {
      it("then the promise is rejected", () => {
         const list = { dataSources: [] };
         adapter.get.mockReturnValueOnce(Promise.resolve(list));
         adapter.map.mockImplementation((response: Promise<any>) => response);

         expect.assertions(1);
         return expect(client.deleteDataSource("Missing data source")).rejects.toEqual("Data source not found: 'Missing data source'");
      });
   });
});
