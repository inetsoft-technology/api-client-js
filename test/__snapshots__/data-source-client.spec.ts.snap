// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataSourceClient when deleteDataSource() is called with a valid name, the force parameter, and succeeds then the promise is resolved 1`] = `
Object {
  "force": "true",
}
`;

exports[`DataSourceClient when getDataSource() is called with a valid name then a data source is returned 1`] = `
Object {
  "defaultDatabase": "PUBLIC",
  "driver": "com.example.Driver",
  "id": "ds1",
  "name": "Data source 1",
  "password": "secret",
  "requireLogin": true,
  "type": "jdbc",
  "url": "jdbc:test:ds1",
  "user": "sa",
}
`;

exports[`DataSourceClient when getDataSources() is called then a list of data sources is returned 1`] = `
Array [
  Object {
    "defaultDatabase": "PUBLIC",
    "driver": "com.example.Driver",
    "id": "ds1",
    "name": "Data source 1",
    "password": "secret",
    "requireLogin": true,
    "type": "jdbc",
    "url": "jdbc:test:ds1",
    "user": "sa",
  },
  Object {
    "id": "ds2",
    "method": "get",
    "name": "Data source 2",
    "type": "text",
    "url": "http://localhost/data/example.csv",
  },
]
`;

exports[`DataSourceClient when updateDataSource() is called with a valid name and is successful then the promise is resolved 1`] = `
Object {
  "id": "ds2",
  "method": "get",
  "name": "Data source 2",
  "type": "text",
  "url": "http://localhost/data/example.csv",
}
`;
